matrix/app_server/vision/perception_encoder.py

```diff
@@ -0,0 +1,269 @@
+
+        else:
+            return JSONResponse(
+                {"error": "Data must be in JSON format!"}, status_code=400
+            )
+
+        output = []
+        try:
+            if "video" in request_json:
+                video_path = request_json["video"].get("file_path")
+                if video_path is None:
+                    return JSONResponse(
+                        {"error": "Request must contain 'file_path' field in 'video'."},
+                        status_code=400,
+                    )
+                start: int = request_json["video"].get("start", 0)
+                end: Optional[int] = request_json["video"].get("end", None)
+                mode: SamplingMode = request_json["video"].get(
+                    "mode", SamplingMode.FRAMES
+                )
+                sampling_rate: Optional[float] = request_json["video"].get(
+                    "sampling_rate", None
+                )
+                window_size: Union[int, float] = request_json["video"].get(
+                    "window_size", 1
+                )
+                stride: Union[int, float] = request_json["video"].get("stride", 1)
+
+                dataset = TorchCodecVideoDataset(
+                    video_path=video_path,
+                    start=start,
+                    end=end,
+                    mode=mode,
+                    window_size=window_size,
+                    stride=stride,
+                    sampling_rate=sampling_rate,
+                    preprocess=self.preprocess,
+                )
+
+                batch_size = (
+                    self.batch_size
+                    if window_size <= 1
+                    else int(0.8 * self.batch_size // max(1, int(window_size)))
+                )
+
+                data_loader = DataLoader(
+                    dataset=dataset,
+                    batch_size=batch_size,
+                    shuffle=False,
+                    num_workers=self.num_workers,
+                    pin_memory=self.device == "cuda",
+                    persistent_workers=self.num_workers > 0,
+                    collate_fn=_custom_collate_fn,
+                )
+
+                with torch.no_grad():
+                    for i, batch in enumerate(data_loader):
+                        frames = batch["frames"].to(
+                            self.device, non_blocking=self.device == "cuda"
+                        )
+                        if window_size > 1:
+                            batch_embeddings = self.model.encode_video(frames)
+                        else:
+                            batch_embeddings = self.model.encode_image(frames)
+                        meta = batch["meta"]
+                        output.append(
+                            {
+                                "embeddings": batch_embeddings.cpu().tolist(),
+                                "meta": meta,
+                            }
+                        )
```

Comment:
asyncio.to_threads

Reply:
Similar to optical flow, The dataloader takes care of efficient loading and the task is GPU-bound, so if the GPU is saturated (by using the right batch size), having multiple threads might not add gains, wdyt?